### PR TITLE
Feature/65-Fix-tests-Dialogue-Manager

### PR DESCRIPTION
**Pull request by IKostric**

### CODE REVIEW

**Review by WerLaj — APPROVED** [review]
LGTM but I checked `_get_elicit_or_recommend_dialogue_acts` only briefly. I found it really difficult to understand the logic inside.

**Review by vinaysetty — CHANGES_REQUESTED** [review]
Looks mostly good just nits and one function with lots of conditions, see if it can be simplieifed.

File: moviebot/dialogue_manager/dialogue_act.py [hunk 1]
```diff
@@ -46,9 +49,26 @@ def __str__(self) -> str:
...
+    def __hash__(self) -> int:
+        """Returns the hash of the dialogue act.
+
+        Returns:
+            Hash of the dialogue act.
+        """
+        return hash((self.intent, tuple(self.params)))
+
     def remove_constraint(self, constraint: ItemConstraint) -> None:
         """Removes constraint from the list of parameters."""
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/dialogue_manager/dialogue_act.py:72]: Done

File: moviebot/dialogue_manager/dialogue_policy.py [hunk 2]
```diff
@@ -30,227 +30,286 @@ def __init__(self, ontology: Ontology, isBot: bool, new_user: bool) -> None:
...
-                        ],
-                    )
-                )
+    def _get_restart_dialogue_act(
+        self, dialogue_state: DialogueState, restart: bool = False
+    ) -> List[DialogueAct]:
+        """Generates the restart dialogue act if applicable.
 
+        Args:
+            dialogue_state: Current dialogue state.
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/dialogue_manager/dialogue_policy.py:90]: Done

File: moviebot/dialogue_manager/dialogue_policy.py [hunk 3]
```diff
@@ -30,227 +30,286 @@ def __init__(self, ontology: Ontology, isBot: bool, new_user: bool) -> None:
...
+            user_intents: Set of user intents.
+
+        Returns:
+            List of dialogue acts.
+        """
+        if UserIntents.BYE in user_intents:
+            return [DialogueAct(AgentIntents.BYE, [])]
+        return []
+
+    def _get_elicit_dialogue_act(
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/dialogue_manager/dialogue_policy.py:121]: done

File: moviebot/dialogue_manager/dialogue_policy.py [hunk 4]
```diff
@@ -30,227 +30,286 @@ def __init__(self, ontology: Ontology, isBot: bool, new_user: bool) -> None:
...
+        self,
+        dialogue_state: DialogueState,
+        user_intents: set[UserIntents],
+        slots: List[str],
+    ) -> List[DialogueAct]:
+        if user_intents.intersection(
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/dialogue_manager/dialogue_policy.py:127]: Done

File: moviebot/dialogue_manager/dialogue_policy.py [hunk 5]
```diff
@@ -30,227 +30,286 @@ def __init__(self, ontology: Ontology, isBot: bool, new_user: bool) -> None:
...
+    def _get_elicit_or_recommend_dialogue_acts(
+        self,
+        dialogue_state: DialogueState,
+        user_dact: DialogueAct,
+        slots: List[str],
+    ) -> List[DialogueAct]:
+        """Generates the elicit or recommend dialogue acts if applicable.
+
+        Args:
+            dialogue_state: Current dialogue state.
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/dialogue_manager/dialogue_policy.py:148]: Done

File: moviebot/dialogue_manager/dialogue_policy.py [hunk 6]
```diff
@@ -30,227 +30,286 @@ def __init__(self, ontology: Ontology, isBot: bool, new_user: bool) -> None:
...
 
         Returns:
-            A list of dialogue acts.
+            A list with the welcome dialogue act.
         """
-        agent_dacts = []
-        slots = deepcopy(dialogue_state.agent_requestable)
         if not dialogue_state.last_user_dacts and not restart:
-            agent_dacts.append(
+            return [
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/dialogue_manager/dialogue_policy.py:73]: It is for consistency with other functions.

File: moviebot/dialogue_manager/dialogue_policy.py [hunk 7]
```diff
@@ -30,227 +30,286 @@ def __init__(self, ontology: Ontology, isBot: bool, new_user: bool) -> None:
...
+            return [self._elict_dialogue_act(slots[0])]
+        return []
+
+    def _get_elicit_or_recommend_dialogue_acts(
+        self,
+        dialogue_state: DialogueState,
+        user_dact: DialogueAct,
+        slots: List[str],
+    ) -> List[DialogueAct]:
+        """Generates the elicit or recommend dialogue acts if applicable.
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/dialogue_manager/dialogue_policy.py:145]: I haven't looked too closely into dialogue flow. The focus in this PR is on refactoring the `next_action` function. The functionality stayed the same. We will address the flow in future PRs.

File: moviebot/dialogue_manager/dialogue_policy.py [hunk 8]
```diff
@@ -30,227 +30,286 @@ def __init__(self, ontology: Ontology, isBot: bool, new_user: bool) -> None:
...
+
+        Args:
+            dialogue_state: Current dialogue state.
+
+        Returns:
+            List of dialogue acts.
+        """
+        agent_dacts = []
```
[the remark this answers is not in the document]
Comment (IKostric) [reply moviebot/dialogue_manager/dialogue_policy.py:153]: Refactored further.

**Comment by IKostric** [comment]
This PR deals with extending the dialogue manager tests and refactoring some complex functions. The dialogue flow is left unchanged.

**Pull request merged**